File: Week 5/src/com/hizamakura/array2/Array2d.java
package com.hizamakura.array2;

import java.util.Arrays;
import java.util.Scanner;
import java.util.stream.Collectors;

public class Array2d {
    static Scanner in = new Scanner(System.in);

    public static void main(String[] args) {
        // input n
        int rows = prompt("Input rows", Integer::parseInt);
        int columns = prompt("Input columns", Integer::parseInt);

        var matrix = new int[rows][columns];

        // input values
        for (int i = 0; i < matrix.length; i++) {
            matrix[i] = promptArr("", columns);
        }

        println("Return");
        // return values
        for (var row : matrix) {
            println(join(" ", 3, row));
        }

        println("Magic square");
        var magicSquare = generateMagicSquare(3);
        for(var row: magicSquare) {
            println(join(" ", 3, row));
        }
    }

    public static int[][] generateMagicSquare(int size) {
        var n = size;
        int[][] magicSquare = new int[n][n];

        // Initialize position for 1
        int r = n / 2;
        int c = n - 1;

        // One by one put all values in magic square
        for (int num = 1; num <= n * n; ) {
            if (r == -1 && c == n) //3rd condition
            {
                c = n - 2;
                r = 0;
            } else {
                //1st condition helper if next number
                // goes to out of square's right side
                if (c == n)
                    c = 0;

                //1st condition helper if next number is
                // goes to out of square's upper side
                if (r < 0)
                    r = n - 1;
            }

            //2nd condition
            if (magicSquare[r][c] != 0) {
                c -= 2;
                r++;
                continue;
            } else
                //set number
                magicSquare[r][c] = num++;

            // move upwards right
            c++;
            r--;
        }

        return magicSquare;
    }

    public static String join(int[] array, int pad) {
        return join(", ", pad, array);
    }

    public static String padRight(String s, int n) {
        return String.format("%-" + n + "s", s);
    }

    public static String padLeft(String s, int n) {
        return String.format("%" + n + "s", s);
    }

    public static String join(String delimiter, int pad, int[] array) {
        var sb = new StringBuilder();
        for(int i = 0; i < array.length; i++) {
            sb.append(padLeft(String.valueOf(array[i]), pad));
            if(i != array.length - 1)
                sb.append(delimiter);
        }
        return sb.toString();
    }

    public static <T> T prompt(String ask, Func<String, T> parse) {
        print(ask + ": ");
        return parse.invoke(in.nextLine());
    }

    public static int[] promptArr(String ask, int size) {
        return prompt(ask, s -> {
            var stream = Arrays.stream(s.split(" "));
            var result = stream
                    .limit(size)
                    .mapToInt(Integer::parseInt)
                    .toArray();
            return result;
        });
    }

    public static void print(String str) {
        System.out.print(str);
    }

    public static void println(String str) {
        System.out.println(str);
    }

    interface Func<T, TResult> {
        public TResult invoke(T arg);
    }

    class FuncImpl<T, TResult> implements Func<T, TResult> {
        Func<T, TResult> function;

        public FuncImpl(Func<T, TResult> function) {
            this.function = function;
        }

        @Override
        public TResult invoke(T arg) {
            return this.function.invoke(arg);
        }
    }
}

File: Week 5/src/com/hizamakura/array2/Array2d.kt
package com.hizamakura.array2

// Time: 1m
fun main() {
    /*// input rows
    val rows = prompt("Input rows") { it.toInt() }

    // input columns
    val columns = prompt("Input columns") { it.toInt() }

    // convert to 2d array
    var matrix = Array(rows) { IntArray(columns) }
    for(row in matrix.indices) {
        matrix[row] = promptArr("", columns) { it.toInt() }.toIntArray()
    }

    println("2d array")
    for (row in matrix) {
        println(row.joinToString(" ") { it.toString().padStart(3) })
    }*/

    println("Magic Square")
    val n = prompt("Input size of magic square") { it.toInt() }
    val magicSquare = generateMagicSquare(n)
    for (row in magicSquare) {
        println("${row.joinToString(" ") { it.toString().padStart(5) }} = ${row.sum()}")
    }
}

fun generateMagicSquare(size: Int): Array<IntArray> {
    val magicSquare = Array(size) { IntArray(size) }

    // Initialize position for 1
    var r = size / 2
    var c = size - 1

    // One by one put all values in magic square
    var num = 1
    while (num <= size * size) {
        if (r == -1 && c == size)
        //3rd condition
        {
            c = size - 2
            r = 0
        } else {
            //1st condition helper if next number
            // goes to out of square's right side
            if (c == size)
                c = 0

            //1st condition helper if next number is
            // goes to out of square's upper side
            if (r < 0)
                r = size - 1
        }

        //2nd condition
        if (magicSquare[r][c] != 0) {
            c -= 2
            r++
            continue
        } else
        //set number
            magicSquare[r][c] = num++
3
        // move upwards right
        c++
        r--
    }

    return magicSquare
}

fun <T> prompt(ask: String, parse: (String) -> T): T {
    print("$ask: ")
    return parse(readLine()!!)
}


inline fun <reified T> promptArr(ask: String, size: Int, crossinline parse: (String) -> T): Array<T> {
    return prompt(ask) { s ->
        val arr = s
                .split(" ")
                .map { parse(it) }
                .take(size)

        arr.toTypedArray();
    };
}
